### ISSUE TITLE
minmax result type can be simplified

### ISSUE THREAD

**Issue by purpleP** [bluss/rust-itertools#351]
Have you considered simplifying `minmax` result type into this?

I'll give two options. I'll use haskellish/mathematical pseudocode for brevity.

Your current version
```
data MinMaxResult a = NoElements | OneElement a | MinMax a a
minmax : [] -> MinMaxResult
```
 Firstly it's better design (but not always feasible with weaker type systems) to reduce function domain than to expand its codomain to make it total.
So what you're doing is you're creating sum type to make function total and handle case where input list has no elements. What you would better be doing in Haskell (I'm not sure if there's alternative in rust) is using special type for lists that always contain at least one element. Then your signature would be simplified into

```
data MinMaxResult a = OneElement a | MinMax a a
minmax : NonEmptyList -> MinMaxResult
```
For all I know rust doesn't have such type in standard library, so it might be impractical in rust.

But then we can see second possible simplification. Is it really useful to distinguish a case where list has only one element? Of course if there's one element minimum and maximum would be the same and then the result type can become just
```
minmax: (PartialOrd a) => [a] -> Maybe (a, a)
```
And this seems mathematically sound. Aren't you're doing to much by kind of trying to also return number of elements in iterator with the result? Do one thing and to it well and all that stuff?

**Comment by scottmcm** [comment]
The `-> Maybe (a, a)` version is spelled `.minmax().into_option()`.

The trick here is that Haskell can use the **contraction** structural rule.  The special type is important in Rust for being able to avoid cloning, either because it's not possible or because it's expensive.  (For example, the smallest `String` might be `"a".repeat(1_000_000)`.)  The `.into_option()` has the `Clone` bound for cases where one wants the simpler return type and is willing to pay the clone cost.

As for a "non-empty" type, that may be possible in general, but Itertools works on standard iterators, which shrink to get elements out, which mean that the type cannot guarantee that it's non-empty for non-infinite iterators.  (And `minmax` is not useful on an infinite one.)

TLDR: Rust is not Haskell, so designs things differently.

**Comment by purpleP** [comment]
@scottmcm By avoiding cloning do you mean `MinMxResult` borrows minimum and maximum value or what?

Thanks for such complete explanation. For some reason nI keep forgetting that iterators are similar, but not the same as cons lists.

**Comment by scottmcm** [comment]
@purpleP It doesn't borrow, but the `MinMaxResult::OneElement` variant means that if there's only one element in the input, it can just move that one element into that variant (without needing to clone it to get a second element).  Whereas if the input contains two or more elements, the min and max are carefully defined to always be distinct elements, so the appropriate two elements are moved into the `MinMaxResult::MinMax` variant.
